File: sdk/ts/src/goals.ts
// Goals API - daemon /goals/* endpoints

import type { SisoulClient } from "./client.js";
import type {
  Goal,
  GoalCreateRequest,
  GoalUpdateRequest,
  ListGoalsResponse,
} from "./types.js";

export class GoalsAPI {
  constructor(private readonly client: SisoulClient) {}

  async list(): Promise<Goal[]> {
    const resp = await this.client.get<ListGoalsResponse>("/goals/list");
    return resp.goals ?? [];
  }

  async add(req: GoalCreateRequest): Promise<Goal> {
    if (!req.title) throw new Error("goals.add: title required");
    return this.client.post<Goal>("/goals/add", req);
  }

  async update(req: GoalUpdateRequest): Promise<Goal> {
    if (!req.id) throw new Error("goals.update: id required");
    return this.client.post<Goal>("/goals/update", req);
  }

  async delete(id: string): Promise<void> {
    if (!id) throw new Error("goals.delete: id required");
    await this.client.post<{ ok: boolean }>("/goals/delete", { id });
  }

  /** 进度小步推进辅助. */
  async bumpProgress(id: string, delta: number): Promise<Goal> {
    const goals = await this.list();
    const g = goals.find((x) => x.id === id);
    if (!g) throw new Error(`goal ${id} not found`);
    const next = Math.min(1, Math.max(0, g.progress + delta));
    return this.update({ id, progress: next });
  }
}

File: sdk/rust/src/goals.rs
//! Goals API.

use crate::client::SisoulClient;
use crate::errors::{Result, SisoulError};
use crate::types::{Goal, GoalCreateRequest, GoalUpdateRequest, ListGoalsResponse};

pub struct GoalsAPI<'a> {
    c: &'a SisoulClient,
}

impl<'a> GoalsAPI<'a> {
    pub fn new(c: &'a SisoulClient) -> Self {
        Self { c }
    }

    pub fn list(&self) -> Result<Vec<Goal>> {
        let r: ListGoalsResponse = self.c.get_json("/goals/list", &[], false)?;
        Ok(r.goals)
    }

    pub fn add(&self, req: &GoalCreateRequest) -> Result<Goal> {
        if req.title.is_empty() {
            return Err(SisoulError::InvalidArgument("goals.add: title required".into()));
        }
        self.c.post_json("/goals/add", req, false)
    }

    pub fn update(&self, req: &GoalUpdateRequest) -> Result<Goal> {
        if req.id.is_empty() {
            return Err(SisoulError::InvalidArgument("goals.update: id required".into()));
        }
        self.c.post_json("/goals/update", req, false)
    }

    pub fn delete(&self, id: &str) -> Result<()> {
        if id.is_empty() {
            return Err(SisoulError::InvalidArgument("goals.delete: id required".into()));
        }
        self.c.post_unit("/goals/delete", &serde_json::json!({ "id": id }), false)
    }

    pub fn bump_progress(&self, id: &str, delta: f64) -> Result<Goal> {
        let goals = self.list()?;
        let target = goals
            .iter()
            .find(|g| g.id == id)
            .ok_or_else(|| SisoulError::InvalidArgument(format!("goal {} not found", id)))?;
        let nxt = (target.progress + delta).clamp(0.0, 1.0);
        self.update(&GoalUpdateRequest {
            id: id.to_string(),
            progress: Some(nxt),
            ..Default::default()
        })
    }
}
